lib/galaxy/webapps/galaxy/services/visualizations.py

```diff
@@ -60,3 +97,360 @@ def index(
+                    raise ValidationError(f"{key} must be a string: {str(type(val))}")
+                val = sanitize_html(val)
+            elif key == "dbkey":
+                if not isinstance(val, str):
+                    raise ValidationError(f"{key} must be a string or unicode: {str(type(val))}")
+                val = sanitize_html(val)
+
+            elif key not in valid_but_uneditable_keys:
+                continue
+                # raise AttributeError( 'unknown key: %s' %( str( key ) ) )
+
+            validated_payload[key] = val
+        if isinstance(payload, VisualizationCreatePayload):
+            return VisualizationCreatePayload(**validated_payload)
+        elif isinstance(payload, VisualizationUpdatePayload):
+            return VisualizationUpdatePayload(**validated_payload)
+
+    def _get_visualization(
+        self,
+        trans: ProvidesUserContext,
+        visualization_id: DecodedDatabaseIdField,
+        check_ownership=True,
+        check_accessible=False,
+    ) -> Visualization:
+        """
+        Get a Visualization from the database by id, verifying ownership.
+        """
+        # Load workflow from database
+        try:
+            visualization = trans.sa_session.get(Visualization, visualization_id)
+        except TypeError:
+            visualization = None
+        if not visualization:
+            raise exceptions.ObjectNotFound("Visualization not found")
+        else:
+            return security_check(trans, visualization, check_ownership, check_accessible)
+
+    def _get_visualization_revision_dict(
+        self,
+        revision: VisualizationRevision,
+    ) -> VisualizationRevisionResponse:
+        """
+        Return a set of detailed attributes for a visualization in dictionary form.
+        NOTE: that encoding ids isn't done here should happen at the caller level.
+        """
+        revision_dict = {
+            "model_class": "VisualizationRevision",
+            "id": revision.id,
+            "visualization_id": revision.visualization.id,
+            "title": revision.title,
+            "dbkey": revision.dbkey,
+            "config": revision.config,
+        }
+        return VisualizationRevisionResponse(**revision_dict)
+
+    def _add_visualization_revision(
+        self,
+        trans: ProvidesUserContext,
+        visualization: Visualization,
+        config: Optional[Union[dict, bytes]],
+        title: Optional[str],
+        dbkey: Optional[str],
+    ) -> VisualizationRevision:
+        """
+        Adds a new `VisualizationRevision` to the given `visualization` with
+        the given parameters and set its parent visualization's `latest_revision`
+        to the new revision.
+        """
+        # precondition: only add new revision on owned vis's
+        # TODO:?? should we default title, dbkey, config? to which: visualization or latest_revision?
+        revision = trans.model.VisualizationRevision(
```

Comment:
I know it was like this before, but I think is better to directly import the models than use `trans.model` to reference them.